### PR TITLE
[Core][VLM] Add support for prefix caching for multi-modal models

### LINKED ISSUE #9790: [Usage]: prefix caching support for multimodal models
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

Hi I want to perform offline batch inference with prefix caching to accelerate inference.
But It looks like it is not supported yet when I initialize LLM with enable_prefix_caching=True

`--enable-prefix-caching is currently not supported for multimodal models and has been disabled.
`
Would it be supported in near future and/or can I get some reference on how to implement it myself?
Currently I'm trying to make batch inference with prefix cache work using huggingface transformers, but it looks like I can't get padding and/or cache position and/or attention_mask right so I get gibberish outputs.

ref:
https://docs.vllm.ai/en/stable/getting_started/examples/offline_inference_with_prefix.html
https://docs.vllm.ai/en/stable/getting_started/examples/offline_inference_vision_language.html
https://github.com/huggingface/transformers/issues/34232#issuecomment-2425907897

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by petersalas**
This adds support for prefix caching with multi-modal models -- in particular it enables it for Ultravox which uses the precise placeholders added in #8346.

With this change, `SelfAttnBlockSpaceManager` et al. now pass a `TokenIds` type around instead of `List[int]` to represent token ids. This new type can also contain `TokenRangeAnnotation`s which capture the contents that will ultimately replace the placeholder tokens. The `Sequence` calculates these by hashing multi-modal content that supports it (currently only implemented for audio).

FIX #9790

### CODE REVIEW

File: vllm/core/block/token_ids.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+from typing import Iterable, List, NamedTuple, Optional, Tuple, Union, overload
```
Comment (petersalas) [inline vllm/core/block/token_ids.py:1]: TODO: add test coverage

File: vllm/core/block/token_ids.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
+from typing import Iterable, List, NamedTuple, Optional, Tuple, overload
+
+from vllm.sequence import Sequence
+
+
+class TokenRangeAnnotation(NamedTuple):
```
Comment (petersalas) [inline vllm/core/block/token_ids.py:6]: Very open to suggestions on naming! This is a pretty bulky name.

File: vllm/core/block_manager_v2.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 from vllm.core.block.interfaces import Block
 from vllm.core.block.prefix_caching_block import (ComputedBlocksTracker,
                                                   LastAccessBlocksTracker)
+from vllm.core.block.token_ids import TokenIds
```
Comment (petersalas) [inline vllm/core/block_manager_v2.py:12]: I only integrated with V2 under the presumption that V2 is largely/nearly stable, but I can look into doing something similar with V1 if warranted.

**Comment by ywang96** [comment]
Sorry for the delay - I was busy with Pixtral release last week but will review this PR this week!

File: vllm/inputs/data.py [hunk 4]
```diff
@@ -122,6 +123,12 @@ class LLMInputs(TypedDict):
     if the model supports it.
     """
 
+    token_annotations: NotRequired[Optional[List["TokenRangeAnnotation"]]]
+    """
+    Optional token annotations to capture content that will replace portions
+    of the token IDs list.
+    """
+
```
Comment (petersalas) [inline vllm/inputs/data.py:131]: Given how #8346 has evolved (placeholder ranges are now propagated instead of inlined into MM data) I'll likely remove this and instead compute the annotations downstream once that change lands. But the rest of the change should still be applicable :)

File: vllm/core/block/token_ids.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    """
+
+    content_hash: int
+    content_offset: int
+    token_index: int
+    token_count: int
+
+    @property
+    def end_index(self) -> int:
+        return self.token_index + self.token_count
```
Comment (Isotr0py) [inline vllm/core/block/token_ids.py:19]: Using `token_index` and `token_count` to represent the "range" is a little bit difficult for reading and comprehension. 

I think using `token_start_index` and `token_end_index` could be better comprehensive, since the `token_count` is fixed for a "range".

Comment (Isotr0py) [reply vllm/core/block/token_ids.py:19]:
```suggestion
class TokenRangeAnnotation(NamedTuple):
    """
    Annotates a range of placeholder tokens to capture content that will
    replace them.
    """

    content_hash: int
    content_offset: int
    token_start_index: int
    token_end_index: int

    @property
    def token_count(self) -> int:
        return self.token_end_index - self.token_start_index
```

File: vllm/core/block_manager_v2.py (hunk 3 above)
Comment (Isotr0py) [reply vllm/core/block_manager_v2.py:12]: Since V1 is deprecating (#8704), I think it's OK to only implement on V2.

File: vllm/core/block/token_ids.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        """
+        Indicates whether two annotations represent adjacent ranges in the
+        hashed content.
+        """
+
+        return (left.content_hash == right.content_hash and
+                left.content_offset + left.token_count == right.content_offset)
+
+    def adjusted(self, tokens_start: int,
+                 tokens_end: int) -> Optional["TokenRangeAnnotation"]:
```
Comment (Isotr0py) [inline vllm/core/block/token_ids.py:33]: I think using `clip` would be better than `adjusted` here for a "range". WDYT?

File: vllm/core/block/token_ids.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+            if key.step:
+                raise IndexError("Step is not supported.")
+
+            # Resolve negative indices.
+            if key.start is None:
+                start = 0
+            elif key.start < 0:
+                start = len(self) + key.start
+            else:
+                start = key.start
```
Comment (Isotr0py) [inline vllm/core/block/token_ids.py:221]:
```suggestion
            start = key.start or 0
            start += len(self) if start < 0 else 0
```

File: vllm/core/block/token_ids.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+            if key.stop is None:
+                stop = len(self)
+            elif key.stop < 0:
+                stop = len(self) + key.stop
+            else:
+                stop = key.stop
```
Comment (Isotr0py) [inline vllm/core/block/token_ids.py:228]:
```suggestion
            stop = key.stop if key.stop is not None else len(self)
            stop += len(self) if stop < 0 else 0
```

File: vllm/inputs/data.py (hunk 4 above)
Comment (afeldman-nm) [reply vllm/inputs/data.py:131]: Just want to clarify, what is the difference between the placeholder range data structures in #8346 and the `TokenRangeAnnotation` data structures in this PR?

Given that you are not creating any new example scripts in this PR, am I correct that the placeholder range data structures in #8346 are more "frontend-oriented" and serve to align placeholder tokens with multimodal input within the prompt (in a way that is model- and workload-specific), while the TokenRangeAnnotation data structures in this PR are more "backend-oriented" and serve to bridge multimodal data into core engine functionality such as prefix cache, block management, etc? With the idea being that the TokenRangeAnnotation's will be computed from the placeholder token range data structures?

File: vllm/core/block/token_ids.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+from vllm.sequence import Sequence
+
+
+class TokenRangeAnnotation(NamedTuple):
+    """
+    Annotates a range of placeholder tokens to capture content that will
+    replace them.
+    """
+
+    content_hash: int
```
Comment (afeldman-nm) [inline vllm/core/block/token_ids.py:12]: This might be a failure on my end, but where are these hashes actually computed? Are these hashes derived from the unprocessed multimodal data (i.e. raw image pixels for images)?

Will there need to be/is there already a way for the engine to automatically choose the appropriate hash function for a given modality?

Are all of these questions contingent on how #8346 gets integrated with this PR?

**Review by afeldman-nm — CHANGES_REQUESTED** [review]
Hi @petersalas had a few nits and some clarifying questions. Overall very excited for this - very cool how multimodal is integrated into prefix caching. Thanks for the PR!

File: vllm/model_executor/models/interfaces.py [hunk 10]
```diff
@@ -27,6 +27,16 @@ class SupportsMultiModal(Protocol):
...
+    supports_chunked_prefill: ClassVar[bool] = False
+    """
+    A flag that indicates this model supports chunked prefill.
+    """
+
+    supports_prefix_caching: ClassVar[bool] = False
+    """
+    A flag that indicates this model supports prefix caching.
+    """
+
```
Comment (petersalas) [inline vllm/model_executor/models/interfaces.py:39]: It's a little weird that these are on `SupportsMultiModal` but the alternative that came to mind was to require tagging every non-multi-modal model as well. Happy to do whatever reviewers think is best here :)

File: vllm/core/block/token_ids.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        return TokenRangeAnnotation(content_hash=self.content_hash,
+                                    content_offset=self.content_offset +
+                                    annotation_start -
+                                    unclamped_annotation_start,
+                                    token_start_index=annotation_start,
+                                    token_end_index=annotation_end)
+
+
+class TokenIds:
+    token_ids: Tuple[int, ...]
```
Comment (petersalas) [inline vllm/core/block/token_ids.py:82]: I don't particularly love `TokenIds.token_ids`. Maybe the type should just be`Tokens`? Maybe `BlockTokens`?

File: vllm/core/block/token_ids.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        Indicates whether two annotations represent adjacent ranges in the
+        hashed content.
+        """
+
+        return (left.content_hash == right.content_hash and
+                left.content_offset + left.token_count == right.content_offset)
+
+    def adjusted(self, tokens_start: int,
+                 tokens_end: int) -> Optional["TokenRangeAnnotation"]:
+        """
```
[the remark this answers is not in the document]
Comment (petersalas) [reply vllm/core/block/token_ids.py:32]: Good suggestion! I added some examples in the docstring which hopefully clarify things a bit.

File: vllm/core/block/token_ids.py (hunk 9 above)
Comment (petersalas) [reply vllm/core/block/token_ids.py:10]: Yup, they were -- originally #8346 took an approach of hiding more multi-modal logic away in each multi-modal model and I was going to do the same for hashing (i.e. delegate it to the model). But since I ended up propagating the placeholder ranges explicitly to the `Sequence` I updated this change to do the MM hashing there as well.

File: vllm/core/block/prefix_caching_block.py [hunk 13]
```diff
@@ -852,7 +856,9 @@ def hash_block_tokens(is_first_block: bool, prev_block_hash: Optional[int],
         - int: The computed hash value for the block.
         """
         assert (prev_block_hash is None) == is_first_block
-        return hash((is_first_block, prev_block_hash, *cur_block_token_ids))
+        return hash(
```
[the remark this answers is not in the document]
Comment (petersalas) [reply vllm/core/block/prefix_caching_block.py:846]: You got it! (With one nit w.r.t. your last question: the hashes are specifically computed for anything that _can't_ be mapped to tokens.)

File: vllm/inputs/data.py (hunk 4 above)
Comment (petersalas) [reply vllm/inputs/data.py]: Yup, that frontend/backend distinction was how I was thinking about it, but I could be convinced to combine them too.

File: vllm/core/block/token_ids.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+            current_token_index = annotation.end_index
+
+        if current_token_index > len(self.token_ids):
+            raise ValueError("TokenRangeAnnotations must be entirely "
+                             "contained within the token IDs.")
+
+    @staticmethod
+    def from_sequence(sequence: Sequence, offset: int) -> "TokenIds":
+        # Since the block table is append-only, the unseen token ids are the
```
[the remark this answers is not in the document]
Comment (petersalas) [reply vllm/core/block/token_ids.py]: Yeah I copy/pasted this comment as part of moving some code around but it didn't really make sense here -- fixed.

File: vllm/core/block/token_ids.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+        # Adjust any annotations for the new token ids list.
+        adjusted_annotations = (a.adjusted(offset, offset + len(token_ids))
+                                for a in sequence.token_annotations)
+        filtered_annotations = (a for a in adjusted_annotations
+                                if a is not None)
+        sorted_annotations = sorted(filtered_annotations,
+                                    key=lambda a: a.token_index)
+        return TokenIds(token_ids, sorted_annotations)
+
+    def chunks(self,
```
[the remark this answers is not in the document]
Comment (petersalas) [reply vllm/core/block/token_ids.py]: 1. Good suggestion -- renamed it to `to_chunks`.
2. I think it'd be tricky to cache, but I did add a fast path in the slice operation since the structure gets sliced for each decoded token and those will always be after the last annotation (if there is any).

**Comment by cooleel** [comment]
Hi, thanks for the great workt! I was wondering if there’s any update on its status or an estimated timeline for its review/merge?

**Comment by ywang96** [comment]
@cooleel We decided to work on adding prefix caching for multimodal models on V1 instead since there are some fundamental changes on how cache manager is designed. Stay tuned!

**Comment by DarkLight1337** [comment]
Closing as superseded by #11187
